zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/store/OpensearchBatchRequest.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.exporter.store;
+
+import io.camunda.exporter.exceptions.OpensearchExporterException;
+import io.camunda.exporter.exceptions.PersistenceException;
+import io.camunda.exporter.utils.OpensearchScriptBuilder;
+import io.camunda.webapps.schema.entities.ExporterEntity;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.client.opensearch._types.OpenSearchException;
+import org.opensearch.client.opensearch._types.Refresh;
+import org.opensearch.client.opensearch.core.BulkRequest;
+import org.opensearch.client.opensearch.core.BulkResponse;
+import org.opensearch.client.opensearch.core.bulk.BulkResponseItem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("rawtypes")
+public class OpensearchBatchRequest implements BatchRequest {
+  public static final int UPDATE_RETRY_COUNT = 3;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OpensearchBatchRequest.class);
+  private final OpenSearchClient osClient;
+  private final BulkRequest.Builder bulkRequestBuilder;
+  private final OpensearchScriptBuilder scriptBuilder;
+
+  public OpensearchBatchRequest(
+      final OpenSearchClient osClient,
+      final BulkRequest.Builder bulkRequestBuilder,
+      final OpensearchScriptBuilder scriptBuilder) {
+    this.osClient = osClient;
+    this.bulkRequestBuilder = bulkRequestBuilder;
+    this.scriptBuilder = scriptBuilder;
+  }
+
+  @Override
+  public BatchRequest add(final String index, final ExporterEntity entity) {
```

Comment:
I don't know how I feel about having this add when `addWithId` exists, ideally there should be as little interface methods as possible. I would prefer `add` to be removed and `addWithId` become `add`.

To be honest I would prefer you have a single `add` function and you pass in some object you built with a builder 
e.g.
`var BatchItem item = new BatchItem.Builder().index(...).id(...).entity(...).routing(...)`
and then you can have a single `add(BatchItem item)` where you can do a single `bulkRequestBuilder.operations`

Reply:
so about this interface in general, I didn't invent it, it was there, and I'm not completely aware of any historic reasons it was build this way for.

I agree with your sentiment in general, but I also don't want to change too many things at once. 

Also, in the bulk operations builders, they have `index`, `update`, `create` and so on, and those should be separated.
You could say the passed object could also contain an operation, yet then I have to keep a list of valid operation, decide upon them and use different methods for different operations. which might in the end make the interface less cluttered, yet the code itself might be more complex.

again, I agree with your sentiment, and I think your idea is worth considering, but I'd say let's see more use cases when more handlers and entities are ported here first and then decide for that

Reply:
ok sure we can wait and see if this changes in the future